### DISCUSSION TITLE
Returning repositories head object in a function

### DISCUSSION THREAD

**Discussion by da-moon** [GitoxideLabs/gitoxide#1112] [Q&A]
Hello!
I have a newb question. I have been trying to write a function that would return my repositories' `HEAD` (`gix::Head`) . I can't figure out how to deal with the borrow checker though as it is returning the following error

```console
error[E0515]: cannot return value referencing local variable `r`
  --> workspaces/version/src/lib.rs:28:5
   |
27 |     let h = r.head()?.to_owned();
   |             - `r` is borrowed here
28 |     Ok(h)
   |     ^^^^^ returns a value referencing data owned by the current function
```

Here is the function :

```rust
fn head<'a>() -> Result<gix::Head<'a>, Box<dyn Error>> {
    let r = gix::discover(Path::new(""))?;
    let r = Box::new(r);
    let h = r.head()?;
    Ok(h)
}
```

Can someone please tell me how I can address this problem ?
Thank you

**Comment by Byron** [comment]
`gix::Head` is actually `gix::Head<'repo>`, which means it borrows the repository that is created a couple of lines earlier, and which is dropped at the end of the function. There are lot's of [great learning resources for Rust](https://www.rust-lang.org/learn) that help understand that borrow checker, and I strongly recommend giving these a shot.

Writing a function like you want it probably doesn't make much sense, but for completeness, here it is:

```Rust
fn head(repo: &gix::Repository) -> Result<gix::Head<'_>, Box<dyn Error>> {
    Ok(r.head()?)
}
```

**Reply by da-moon** [reply]
so correct me if I am wrong but there is no proper way to create the repo inside the function, right? I wanted to keep the function signature as something like `fn head() -> Result<gix::Head<'_>, Box<dyn Error>>`  and not pass in a reference to the repo object that was created out of the function's scope.

I tried wrapping the Repository object and Head in `Rc` and `Box` but none of them worked.
